source/distributed/solution_transfer.cc

```diff
@@ -247,27 +341,90 @@ namespace parallel
       const std::vector<::dealii::Vector<typename VectorType::value_type>>
         dofvalues = Utilities::unpack<
           std::vector<::dealii::Vector<typename VectorType::value_type>>>(
-          data_range.begin(), data_range.end(), /*allow_compression=*/false);
+          data_range.begin(), data_range.end(), has_hp_dof_handler);
 
       // check if sizes match
       Assert(dofvalues.size() == all_out.size(), ExcInternalError());
 
-      // check if we have enough dofs provided by the FE object
-      // to interpolate the transferred data correctly
-      for (auto it_dofvalues = dofvalues.begin();
-           it_dofvalues != dofvalues.end();
-           ++it_dofvalues)
-        Assert(
-          cell->get_fe().dofs_per_cell == it_dofvalues->size(),
-          ExcMessage(
-            "The transferred data was packed with a different number of dofs than the"
-            "currently registered FE object assigned to the DoFHandler has."));
-
-      // distribute data for each registered vector on mesh
-      auto it_input  = dofvalues.cbegin();
-      auto it_output = all_out.begin();
-      for (; it_input != dofvalues.cend(); ++it_input, ++it_output)
-        cell->set_dof_values_by_interpolation(*it_input, *(*it_output));
+      if (has_hp_dof_handler)
+        {
+          unsigned int fe_index = numbers::invalid_unsigned_int;
+
+          switch (status)
+            {
+              case parallel::distributed::Triangulation<
+                dim,
+                DoFHandlerType::space_dimension>::CELL_PERSIST:
+              case parallel::distributed::Triangulation<
+                dim,
+                DoFHandlerType::space_dimension>::CELL_COARSEN:
+                fe_index = cell->active_fe_index();
+                break;
+
+              case parallel::distributed::Triangulation<
+                dim,
+                DoFHandlerType::space_dimension>::CELL_REFINE:
+                // set fe index to the one of the first child,
+                // and check if all children have the same fe index
+                fe_index = cell->child(0)->active_fe_index();
+                for (unsigned int child_index = 1;
+                     child_index < GeometryInfo<dim>::max_children_per_cell;
+                     ++child_index)
+                  Assert(cell->child(child_index)->active_fe_index() ==
+                           fe_index,
+                         ExcInternalError());
```

Comment:
out of curiosity, why does this condition have to be true? maybe just put the explanation into the comment above

Reply:
Or write ExcMessage()?